BlueprintUICommonControls/Sources/ScrollView.swift

```diff
@@ -260,8 +300,137 @@ fileprivate final class ScrollerWrapperView: UIView {
+        }
+    }
+}
 
 
+extension ScrollView
+{
+    // Calculates the correct content inset to apply for the given inputs.
+    
+    static func finalContentInset(
+        scrollViewInsets : UIEdgeInsets,
+        safeAreaInsets : UIEdgeInsets,
+        keyboardBottomInset : CGFloat,
+        refreshControlState : PullToRefreshBehavior,
+        refreshControlBounds : CGRect?
+    ) -> UIEdgeInsets
+    {
+        var finalContentInset = scrollViewInsets
+        
+        // Include the keyboard's adjustment at the bottom of the scroll view.
+        
+        if keyboardBottomInset > 0.0 {
+            finalContentInset.bottom += keyboardBottomInset
+            
+            // Exclude the safe area insets, so the content hugs the top of the keyboard.
+            
+            finalContentInset.bottom -= safeAreaInsets.bottom
+        }
+        
+        // The refresh control lives above the content and adjusts the
+        // content inset for itself when visible and refreshing.
+        // Do the same adjustment to our expected content inset.
+        
+        if case .refreshing = refreshControlState {
+            finalContentInset.top += refreshControlBounds?.size.height ?? 0.0
+        }
+        
+        return finalContentInset
     }
+}
 
+
+extension ScrollerWrapperView : KeyboardObserverDelegate {
+    
+    //
+    // MARK: Keyboard
+    //
+    
+    private func updateBottomContentInsetWithKeyboardFrame() {
+        
+        let contentInset = ScrollView.finalContentInset(
+            scrollViewInsets: self.representedElement.contentInset,
+            safeAreaInsets: self.bp_safeAreaInsets,
+            keyboardBottomInset: self.bottomContentInsetAdjustmentForKeyboard,
+            refreshControlState: self.representedElement.pullToRefreshBehavior,
+            refreshControlBounds: self.refreshControl?.bounds
+        )
+        
+        /// Setting contentInset, even to the same value, can cause issues during scrolling (such as stopping scrolling).
+        /// Make sure we're only assigning the value if it changed.
+        
+        if self.scrollView.contentInset.bottom != contentInset.bottom {
+            self.scrollView.contentInset.bottom = contentInset.bottom
+        }
+    }
+    
+    fileprivate var bottomContentInsetAdjustmentForKeyboard : CGFloat {
+        
+        switch self.representedElement.keyboardAdjustmentMode {
+        case .none:
+            return 0.0
+            
+        case .adjustsWhenVisible:
+            guard let keyboardFrame = self.keyboardObserver.currentFrame(in: self) else {
+                return 0.0
+            }
+            
+            return {
+                switch keyboardFrame {
+                case .nonOverlapping:
+                    return 0.0
+                    
+                case .visible(let frame):
+                    return self.bounds.size.height - frame.origin.y
+                }
+            }()
```

Comment:
I think you can remove this closure wrapper?

Reply:
Oops, yep